R/data.table.R

```diff
@@ -1861,90 +1861,245 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+  }
+  
+  # Next determine if any columns are list or non-atomic type. If so, convert all columns to lists
+  col.types = sapply(X, typeof)
+  uniq.col.types = unique(col.types)
+  atomics <- c("logical", "integer", "numeric", "double", "complex", "character", "raw")
+  any.non.atomic = (length(setdiff(uniq.col.types, atomics)) > 0L)
+  if (any.non.atomic) {
     for (j in seq_len(p)) {
-      xj = X[[j]]
-      if (is.recursive(xj)) { }
-      else X[[j]] = as.list(as.vector(xj))
+      if (!is.recursive(X[[j]])) { 
+        X[[j]] = as.list(as.vector(X[[j]]))
+      }
     }
-  }
-  else if (all.logical) { }
-  else if (non.numeric) {
-    for (j in seq_len(p)) {
-      if (is.character(X[[j]])) next
-      xj = X[[j]]
-      miss = is.na(xj)
-      xj = if (length(levels(xj))) as.vector(xj) else format(xj)
-      is.na(xj) = miss
-      X[[j]] = xj
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Convert factors to character vectors
+  any.factors = ("factor" %chin% uniq.classes)
+  if (any.factors) {
+    which.factors = which(sapply(col.classes, function(cl_vec) { "factor" %chin% cl_vec }))
+    for (j in which.factors) {
+      miss = is.na(X[[j]])
+      X[[j]] = as.vector(X[[j]])
+      is.na(X[[j]]) = miss
+    }
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Classes to be converted to character vectors
+  charconvert.classes = c("Date", "POSIXct", "POSIXlt", "raw")
+  any.charconvert = any(charconvert.classes %chin% uniq.classes)
+  if (any.charconvert) {
+    which.charconvert = which(sapply(col.classes, function(cl_vec) { any(charconvert.classes %chin% cl_vec) }))
+    for (j in which.charconvert) {
+      miss = is.na(X[[j]])
+      X[[j]] = format(X[[j]])
+      is.na(X[[j]]) = miss
+    }
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Other classes from suggested packages that have special type conversion rules
+  any.bit64 = ("integer64" %chin% uniq.classes)
+  if (any.bit64 && length(uniq.classes) > 1L) {
+    # Determine target class we're converting to 
+    target.class = NULL
+    if (!requireNamespace("bit64", quietly=TRUE)) 
+      target.class = "character" # bit64 package not installed
```

Comment:
ditto for `# nocov`

Reply:
> Looking good overall! Thanks for all the work.
> 
> The logic for aligning to an output type is something that should probably be shared in other places (`rbindlist` and ongoing `fexplode` come to mind), I wonder if we can't build something for that to share logic...
> 
> I wouldn't focus on it for this PR though, something to do as a follow-up.

I agree this would be a good idea, I'm finding a few ways to construct malformed data.tables that should be handled before handing off to C code